// react-stormpath/lib/components/RegistrationForm.js
'use strict';

var _dec, _class;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactMixin = require('react-mixin');

var _reactMixin2 = _interopRequireDefault(_reactMixin);

var _reactRouter = require('react-router');

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _LoginLink = require('../components/LoginLink');

var _LoginLink2 = _interopRequireDefault(_LoginLink);

var _UserActions = require('../actions/UserActions');

var _UserActions2 = _interopRequireDefault(_UserActions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DefaultRegistrationForm = function (_React$Component) {
  _inherits(DefaultRegistrationForm, _React$Component);

  function DefaultRegistrationForm() {
    _classCallCheck(this, DefaultRegistrationForm);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(DefaultRegistrationForm).apply(this, arguments));
  }

  _createClass(DefaultRegistrationForm, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        RegistrationForm,
        this.props,
        _react2.default.createElement(
          'div',
          { className: 'sp-login-form' },
          _react2.default.createElement(
            'div',
            { className: 'row', showWhen: 'account.created' },
            _react2.default.createElement(
              'div',
              { className: 'col-sm-offset-4 col-xs-12 col-sm-4' },
              _react2.default.createElement(
                'p',
                { className: 'alert alert-success', showWhen: 'account.enabled' },
                'Your account has been created. ',
                _react2.default.createElement(
                  _LoginLink2.default,
                  null,
                  'Login Now'
                ),
                '.'
              ),
              _react2.default.createElement(
                'div',
                { showWhen: '!account.enabled' },
                _react2.default.createElement(
                  'p',
                  { className: 'alert alert-success' },
                  'Your account has been created. Please check your email for a verification link.'
                ),
                _react2.default.createElement(
                  'p',
                  { className: 'pull-right' },
                  _react2.default.createElement(
                    _LoginLink2.default,
                    null,
                    'Back to Login'
                  )
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'row', showWhen: '!account.created' },
            _react2.default.createElement(
              'div',
              { className: 'col-xs-12' },
              _react2.default.createElement(
                'div',
                { className: 'form-horizontal' },
                _react2.default.createElement(
                  'div',
                  { className: 'form-group' },
                  _react2.default.createElement(
                    'label',
                    { htmlFor: 'spFirstName', className: 'col-xs-12 col-sm-4 control-label' },
                    'First Name'
                  ),
                  _react2.default.createElement(
                    'div',
                    { className: 'col-xs-12 col-sm-4' },
                    _react2.default.createElement('input', { className: 'form-control', id: 'spFirstName', placeholder: 'First Name', name: 'givenName' })
                  )
                ),
                _react2.default.createElement(
                  'div',
                  { className: 'form-group' },
                  _react2.default.createElement(
                    'label',
                    { htmlFor: 'spLastName', className: 'col-xs-12 col-sm-4 control-label' },
                    'Last Name'
                  ),
                  _react2.default.createElement(
                    'div',
                    { className: 'col-xs-12 col-sm-4' },
                    _react2.default.createElement('input', { className: 'form-control', id: 'spLastName', placeholder: 'Last Name', name: 'surname' })
                  )
                ),
                _react2.default.createElement(
                  'div',
                  { className: 'form-group' },
                  _react2.default.createElement(
                    'label',
                    { htmlFor: 'spEmail', className: 'col-xs-12 col-sm-4 control-label' },
                    'Email'
                  ),
                  _react2.default.createElement(
                    'div',
                    { className: 'col-xs-12 col-sm-4' },
                    _react2.default.createElement('input', { className: 'form-control', id: 'spEmail', placeholder: 'Email', name: 'email' })
                  )
                ),
                _react2.default.createElement(
                  'div',
                  { className: 'form-group' },
                  _react2.default.createElement(
                    'label',
                    { htmlFor: 'spPassword', className: 'col-xs-12 col-sm-4 control-label' },
                    'Password'
                  ),
                  _react2.default.createElement(
                    'div',
                    { className: 'col-xs-12 col-sm-4' },
                    _react2.default.createElement('input', { type: 'password', className: 'form-control', id: 'spPassword', placeholder: 'Password', name: 'password' })
                  )
                ),
                _react2.default.createElement(
                  'div',
                  { className: 'form-group' },
                  _react2.default.createElement(
                    'div',
                    { className: 'col-sm-offset-4 col-sm-4' },
                    _react2.default.createElement(
                      'p',
                      { className: 'alert alert-danger', showWhen: 'form.error' },
                      _react2.default.createElement('span', { replaceWith: 'form.errorMessage' })
                    ),
                    _react2.default.createElement(
                      'button',
                      { type: 'submit', className: 'btn btn-primary' },
                      'Register'
                    )
                  )
                )
              )
            )
          )
        )
      );
    }
  }]);

  return DefaultRegistrationForm;
}(_react2.default.Component);

var RegistrationForm = (_dec = _reactMixin2.default.decorate(_reactRouter.History), _dec(_class = function (_React$Component2) {
  _inherits(RegistrationForm, _React$Component2);

  function RegistrationForm() {
    var _Object$getPrototypeO;

    var _temp, _this2, _ret;

    _classCallCheck(this, RegistrationForm);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this2 = _possibleConstructorReturn(this, (_Object$getPrototypeO = Object.getPrototypeOf(RegistrationForm)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this2), _this2.state = {
      fields: {
        givenName: '',
        surname: '',
        email: '',
        password: ''
      },
      errorMessage: null,
      isFormProcessing: false,
      isAccountCreated: false,
      isAccountEnabled: false
    }, _temp), _possibleConstructorReturn(_this2, _ret);
  }

  _createClass(RegistrationForm, [{
    key: 'onFormSubmit',
    value: function onFormSubmit(e) {
      var _this3 = this;

      e.preventDefault();

      var next = function next(err, data) {
        if (err) {
          return _this3.setState({
            isFormProcessing: false,
            errorMessage: err.message
          });
        }

        // If the user didn't specify any data,
        // then simply default to what we have in state.
        data = data || _this3.state.fields;

        _UserActions2.default.register(data, function (err, result) {
          if (err) {
            _this3.setState({
              isFormProcessing: false,
              errorMessage: err.message
            });
          } else if (result.status === 'ENABLED') {
            _UserActions2.default.login({
              username: data.email || data.username,
              password: data.password
            }, function (err) {
              if (err) {
                return _this3.setState({
                  isFormProcessing: false,
                  isAccountCreated: false,
                  errorMessage: err.message
                });
              }

              _this3.setState({
                isFormProcessing: false,
                isAccountCreated: true,
                isAccountEnabled: true
              });
            });
          } else {
            _this3.setState({
              isFormProcessing: false,
              isAccountCreated: true,
              isAccountEnabled: false
            });
          }
        });
      };

      this.setState({
        isFormProcessing: true
      });

      if (this.props.onSubmit) {
        e.data = this.state.fields;
        this.props.onSubmit(e, next);
      } else {
        next(null, this.state.fields);
      }
    }
  }, {
    key: '_mapFormFieldHandler',
    value: function _mapFormFieldHandler(element, tryMapField) {
      if (['input', 'textarea'].indexOf(element.type) > -1) {
        if (element.props.type !== 'submit') {
          switch (element.props.name) {
            case 'email':
              tryMapField('email');
              break;
            case 'login':
            case 'username':
              tryMapField('username');
              break;
            case 'givenName':
            case 'firstName':
              tryMapField('givenName');
              break;
            case 'surname':
            case 'lastName':
              tryMapField('surname');
              break;
            case 'password':
              tryMapField('password');
              break;
          }
        }
      }
    }
  }, {
    key: '_showWhenHandler',
    value: function _showWhenHandler(action, element) {
      var test = null;

      switch (action) {
        case 'form.processing':
          test = this.state.isFormProcessing;
          break;
        case 'form.error':
          test = !!this.state.errorMessage;
          break;
        case 'account.created':
          test = this.state.isAccountCreated;
          break;
        case 'account.enabled':
          test = this.state.isAccountEnabled;
          break;
      }

      return test;
    }
  }, {
    key: '_replaceWithHandler',
    value: function _replaceWithHandler(action, element) {
      var element = false;

      switch (action) {
        case 'form.errorMessage':
          var className = element.props ? element.props.className : undefined;
          element = _react2.default.createElement(
            'span',
            { className: className },
            this.state.errorMessage
          );
          break;
      }

      return element;
    }
  }, {
    key: 'render',
    value: function render() {
      if (this.props.children) {
        return _react2.default.createElement(
          'form',
          { onSubmit: this.onFormSubmit.bind(this) },
          _utils2.default.makeForm(this, this._mapFormFieldHandler.bind(this), this._showWhenHandler.bind(this), this._replaceWithHandler.bind(this))
        );
      } else {
        return _react2.default.createElement(DefaultRegistrationForm, this.props);
      }
    }
  }]);

  return RegistrationForm;
}(_react2.default.Component)) || _class);
exports.default = RegistrationForm;
