// react-stormpath/es6/Context.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var contextInstance = null;

var Context = function () {
  function Context() {
    _classCallCheck(this, Context);

    if (contextInstance) {
      return contextInstance;
    }

    this.router = null;
    this.dispatcher = null;
    this.endpoints = null;

    contextInstance = this;
  }

  _createClass(Context, [{
    key: "setRouter",
    value: function setRouter(router) {
      this.router = router;
    }
  }, {
    key: "getRouter",
    value: function getRouter() {
      return this.router;
    }
  }, {
    key: "setDispatcher",
    value: function setDispatcher(dispatcher) {
      this.dispatcher = dispatcher;
    }
  }, {
    key: "getDispatcher",
    value: function getDispatcher() {
      return this.dispatcher;
    }
  }, {
    key: "setEndpoints",
    value: function setEndpoints(endpoints) {
      this.endpoints = endpoints;
    }
  }, {
    key: "getEndpoints",
    value: function getEndpoints() {
      return this.endpoints || {};
    }
  }], [{
    key: "getInstance",
    value: function getInstance() {
      if (!contextInstance) {
        return new Context();
      }
      return contextInstance;
    }
  }]);

  return Context;
}();

exports.default = Context;
