// pui-react-pagination/pagination.js
/*(c) Copyright 2015 Pivotal Software, Inc. All Rights Reserved.*/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pagination = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _puiReactHelpers = require('pui-react-helpers');

require('pui-css-pagination');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var types = _react2.default.PropTypes;

var PaginationButton = function (_React$Component) {
  (0, _inherits3.default)(PaginationButton, _React$Component);

  function PaginationButton() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PaginationButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PaginationButton.__proto__ || (0, _getPrototypeOf2.default)(PaginationButton)).call.apply(_ref, [this].concat(args))), _this), _this.click = function (e) {
      var _this$props = _this.props,
          eventKey = _this$props.eventKey,
          onSelect = _this$props.onSelect;

      onSelect && onSelect(e, { eventKey: eventKey });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PaginationButton, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          content = _props.content,
          active = _props.active;

      var activeClass = active ? 'active' : '';
      return _react2.default.createElement(
        'li',
        { onClick: this.click, className: activeClass },
        _react2.default.createElement(
          'a',
          null,
          content
        )
      );
    }
  }]);
  return PaginationButton;
}(_react2.default.Component);

PaginationButton.propTypes = {
  content: types.node,
  active: types.bool,
  onSelect: types.func,
  eventKey: types.oneOfType([types.number, types.string])
};

var Pagination = exports.Pagination = function (_React$Component2) {
  (0, _inherits3.default)(Pagination, _React$Component2);

  function Pagination() {
    (0, _classCallCheck3.default)(this, Pagination);
    return (0, _possibleConstructorReturn3.default)(this, (Pagination.__proto__ || (0, _getPrototypeOf2.default)(Pagination)).apply(this, arguments));
  }

  (0, _createClass3.default)(Pagination, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          items = _props2.items,
          next = _props2.next,
          prev = _props2.prev,
          activePage = _props2.activePage,
          onSelect = _props2.onSelect,
          props = (0, _objectWithoutProperties3.default)(_props2, ['items', 'next', 'prev', 'activePage', 'onSelect']);

      var paginationButtons = [].concat((0, _toConsumableArray3.default)(Array(items))).map(function (_, index) {
        var isActive = index + 1 === activePage;
        return _react2.default.createElement(PaginationButton, (0, _extends3.default)({
          key: index,
          content: index + 1,
          active: isActive,
          onSelect: onSelect,
          eventKey: index + 1
        }, props));
      });

      var prevButton = _react2.default.createElement(PaginationButton, { onSelect: onSelect, eventKey: 'prev', content: '\u2039' });
      var nextButton = _react2.default.createElement(PaginationButton, { onSelect: onSelect, eventKey: 'next', content: '\u203A' });

      return _react2.default.createElement(
        'ul',
        { className: 'pagination' },
        prev ? prevButton : null,
        paginationButtons,
        next ? nextButton : null
      );
    }
  }]);
  return Pagination;
}(_react2.default.Component);

Pagination.propTypes = {
  items: types.number,
  next: types.bool,
  prev: types.bool,
  activePage: types.number,
  onSelect: types.func
};
Pagination.defaultProps = {
  next: true,
  prev: true,
  onSelect: function onSelect() {}
};
